src/lib.rs

```diff
@@ -0,0 +1,534 @@
+                        let mut timeout_checker = timeout_checker.clone();
+                        let handle = scope.spawn(move || unsafe {
+                            test(chunk.as_mut_ptr(), chunk.len(), &mut timeout_checker)
+                        });
+                        handles.push(handle);
+                    }
+
+                    handles
+                        .into_iter()
+                        .map(|handle| {
+                            handle
+                                .join()
+                                .unwrap_or(Err(MemtestError::Other(anyhow!("Thread panicked"))))
+                        })
+                        .fold(Ok(MemtestOutcome::Pass), |acc, result| {
+                            use {MemtestError::*, MemtestOutcome::*};
+                            match (acc, result) {
+                                (Err(Other(e)), _) | (_, Err(Other(e))) => Err(Other(e)),
+                                (Err(Timeout), _) | (_, Err(Timeout)) => Err(Timeout),
+                                (Ok(Fail(f)), _) | (_, Ok(Fail(f))) => Ok(Fail(f)),
+                                _ => Ok(Pass),
+                            }
+                        })
+                })
+            } else {
+                unsafe { test(memory.as_mut_ptr(), memory.len(), timeout_checker) }
+            };
+
+            if matches!(test_result, Ok(MemtestOutcome::Fail(_))) && self.allow_early_termination {
+                reports.push(MemtestReport::new(*test_type, test_result));
+                warn!("Memtest failed, terminating early");
+                break;
+            }
+            reports.push(MemtestReport::new(*test_type, test_result));
+        }
+
+        Ok(reports)
+    }
+}
+
+impl fmt::Display for MemtestReportList {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "tested_memsize = {}", self.tested_usize_count)?;
+        writeln!(f, "mlocked = {}", self.mlocked)?;
+        for report in &self.reports {
+            let outcome = match &report.outcome {
+                Ok(outcome) => format!("{}", outcome),
+                Err(e) => format!("{}", e),
+            };
+            writeln!(
+                f,
+                "{:<30} {}",
+                format!("Ran {:?}", report.test_type),
+                outcome
+            )?;
+        }
+        Ok(())
+    }
+}
+
+impl MemtestReportList {
+    pub fn iter(&self) -> std::slice::Iter<'_, MemtestReport> {
+        self.reports.iter()
+    }
+
+    /// Returns true if all tests were run successfully and all tests passed
+    pub fn all_pass(&self) -> bool {
+        self.iter()
+            .all(|report| matches!(report.outcome, Ok(MemtestOutcome::Pass)))
+    }
+}
+
+impl MemtestReport {
+    fn new(test_type: MemtestType, outcome: Result<MemtestOutcome, MemtestError>) -> MemtestReport {
+        MemtestReport { test_type, outcome }
+    }
+}
+
+impl TimeoutChecker {
```

Comment:
Alright, so... I went a bit off the rails thinking about how we could do this in a way that I hope makes a bit more sense, and here's what I came up with: Have the checker calculate how much time is left before the deadline, schedule the next check at 75% of that interval, and then estimate the number of iterations to get there. And repeat that every check.

It has the effect of making it so that the rate of checking gets faster the closer you get to the deadline. Let's say you have 1000ms left -- Your next check will be at 75% of that (approx 750ms). The next check, you have roughly 250ms left, so your next check will be 75% (approx 187ms). Then you have 63ms left -- 75% of that means you check again in 47ms, etc etc. Eventually you will check in microseconds and then nanoseconds, etc. (Of course, subject to the granularity of your test iterations).

Here is the code:

```Rust
#[derive(Clone, Debug)]
struct TimeoutChecker {
    deadline: Instant,
    curr_test: Option<TimeoutCheckerTestInstance>,
}

#[derive(Clone, Debug)]
struct TimeoutCheckerTestInstance {
    test_start_time: Instant,
    expected_iter: u64,
    completed_iter: u64,
    checkpoint: u64,
    num_checks_completed: u64,
    last_progress_fraction: f32,
}

impl TimeoutChecker {
    fn new(timeout: Duration) -> TimeoutChecker {
        TimeoutChecker {
            deadline: Instant::now() + timeout,
            curr_test: None,
        }
    }
    fn init(&mut self, expected_iter: u64) {
        self.curr_test = Some(TimeoutCheckerTestInstance {
            test_start_time: Instant::now(),
            expected_iter,
            completed_iter: 0,
            checkpoint: 1,
            num_checks_completed: 0,
            last_progress_fraction: 0.0,
        });
    }
    fn check(&mut self) -> Result<(), MemtestError> {
        const DEADLINE_CHECK_RATIO: f64 = 0.75;

        let Some(test_instance) = &mut self.curr_test else {
            panic!("Attempted to check timeout without initialization");
        };

        let progress_fraction =
            test_instance.completed_iter as f32 / test_instance.expected_iter as f32;
        if progress_fraction - test_instance.last_progress_fraction >= 0.01 {
            trace!("Progress: {:.0}%", progress_fraction * 100.0);
            test_instance.last_progress_fraction = progress_fraction;
        }

        if test_instance.completed_iter < test_instance.checkpoint {
            test_instance.completed_iter += 1;
            return Ok(());
        }

        let current_time = Instant::now();
        if current_time >= self.deadline {
            return Err(MemtestError::Timeout);
        }

        let duration_until_next_checkpoint = {
            let duration_until_deadline = self.deadline - current_time;
            duration_until_deadline.mul_f64(DEADLINE_CHECK_RATIO)
        };

        let avg_iter_duration = {
            let test_elapsed = current_time - test_instance.test_start_time;
            test_elapsed.div_f64(test_instance.completed_iter as f64)
        };

        let iter_until_next_checkpoint = {
            let x = duration_until_next_checkpoint.div_duration_f64(avg_iter_duration) as u64;
            u64::max(x, 1)
        };

        test_instance.checkpoint += iter_until_next_checkpoint;

        test_instance.num_checks_completed += 1;
        test_instance.completed_iter += 1;
        Ok(())
    }
}
```